File: Cpp/0706_design_hashmap.cpp
// https://leetcode.cn/problems/design-hashmap/

class MyHashMap {
public:
    MyHashMap() {
        size_ = 1000;
        data_.resize(size_);
    }
    
    void put(int key, int value) {
        int idx = hash(key);
        for (auto it = data_[idx].begin(); it != data_[idx].end(); ++it) {
            if (it->first == key) {
                it->second = value;
                return;
            }
        }
        data_[idx].push_back(make_pair(key, value));
    }
    
    int get(int key) {
        int idx = hash(key);
        for (auto it = data_[idx].begin(); it != data_[idx].end(); ++it) {
            if (it->first == key) {
                return it->second;
            }
        }
        return -1;
    }
    
    void remove(int key) {
        int idx = hash(key);
        for (auto it = data_[idx].begin(); it != data_[idx].end(); ++it) {
            if (it->first == key) {
                data_[idx].erase(it);
                return;
            }
        }
    }

    int hash(int key) {
        return key % size_;
    }

    vector<list<pair<int,int>>> data_;
    int size_;
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */

File: Python3/0706_design_hashmap.py
# https://leetcode.cn/problems/design-hashmap/

class MyHashMap:

    def __init__(self):
        self.size = 1000
        self.data = [[] for _ in range(self.size)]

    def put(self, key: int, value: int) -> None:
        idx = self.hash(key)
        for i in range(len(self.data[idx])):
            if self.data[idx][i][0] == key:
                self.data[idx][i][1] = value
                return
        self.data[idx].append([key, value])

    def get(self, key: int) -> int:
        idx = self.hash(key)
        for i in range(len(self.data[idx])):
            if self.data[idx][i][0] == key:
                return self.data[idx][i][1]
        return -1

    def remove(self, key: int) -> None:
        idx = self.hash(key)
        for i in range(len(self.data[idx])):
            if self.data[idx][i][0] == key:
                value = self.data[idx][i][1]
                self.data[idx].remove([key, value])
                return

    def hash(self, key):
        return key % self.size



# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)